### PR TITLE
Update Run statuses correctly and stop reconciling invalid Runs

### LINKED ISSUE #174: Runs should stop reconciling if they cannot be run
## Expected Behavior
If a `TaskRun` or a `PipelineRun` runs into an error that prevents it from being able to run, it should not be reconciled and the reconciler/controller should stop trying to reconcile it.

The status/conditions of the Run should also be updated to indicate that the Run has failed.

## Actual Behavior
In most cases, e.g. when a TaskRun references a Task that doesn't exist, the reconciler will continue trying to reconcile, and the status/conditions will remain empty.

This means that if the referenced Task or Pipeline is created later on, suddenly the Run will start to execute.

## Steps to Reproduce the Problem

  1. Create a TaskRun that references a Task that doesn't exist (or do the same with a PipelineRun + Pipeline)
  2. Use `kubectl get taskruns -o yaml` and see that the status/conditions are empty
  3. View the logs for the controller (`kubectl -n knative-build-pipeline logs $(kubectl -n knative-build-pipeline get pods -l app=build-pipeline-controller -o name)`) and see that there is an error b/c the referenced object does not exist
4. Create the referenced Task or Pipeline
5. Notice (`kubectl get taskruns`) that suddenly the Run is running

### PR DESCRIPTION
**Pull request by bobcatfish**
If a Run fails to start, the status should indicate it has failed.
If it is currently running, the status should indicate it is in-progress.

Otherwise, we can assume the related Builds were created and we mirror
their statuses.

Changed the log level for invalid Runs to Info b/c this means the
_user_ made an error, the Error log level should be used for when there
is an Error in the system/controller itself.

Includes some refactoring of the controller test libs. In the long run, I think we should be moving functionality out of the reconcilers themselves so that we can test it more easily without having to use fake informers/clients, and without having to assert on logs.

Fixes #174

### CODE REVIEW

**Comment by shashwathi** [comment]
@bobcatfish : It would be more useful to propagate meaningful msg in status to explain the failure instead of just setting the status to false. For now it could be the error msg too.

WDYT?

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 1]
```diff
@@ -124,14 +125,23 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
 		return nil
 	} else if err != nil {
+		c.Logger.Errorf("Error retreiving TaskRun %q: %s", name, err)
 		return err
 	}
 
 	// Don't modify the informer's copy.
 	tr := original.DeepCopy()
+	// If this is the first attempt to reconcile, mark the Run as in progress
+	if tr.Status.GetCondition(duckv1alpha1.ConditionSucceeded) == nil {
```
Comment (shashwathi) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: I would probably make this like `InitizlieStatusCondition()` function and move the check for nil and setting condition inside that.

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 2]
```diff
@@ -152,17 +154,27 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) error {
 	p, err := c.pipelineLister.Pipelines(pr.Namespace).Get(pr.Spec.PipelineRef.Name)
```
Comment (shashwathi) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: For consistency across taskrun and pipelinerun reconcilers, I would suggest using similar `InitializeConditions()` function at the beginning of `reconcile` function.

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 3]
```diff
@@ -179,10 +191,14 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
...
+			pr.Status.SetCondition(&duckv1alpha1.Condition{
+				Type:   duckv1alpha1.ConditionSucceeded,
+				Status: corev1.ConditionFalse,
+			})
 			return nil
 		}
-		return fmt.Errorf("error getting Tasks and/or TaskRuns for Pipeline %s, Pipeline may be invalid!: %s", p.Name, err)
+		return fmt.Errorf("error getting Tasks and/or TaskRuns for Pipeline %s: %s", p.Name, err)
 	}
 	prtr := resources.GetNextTask(pr.Name, state, c.Logger)
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: If there are no errors, set the condition to `Unknown` to indicate pipeline run is still running. Can also be done in the beginning of `reconcile()` before checking for errors

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 4]
```diff
@@ -393,77 +395,11 @@ func TestReconcileBuildUpdateStatus(t *testing.T) {
 		t.Fatalf("Unexpected error when Reconcile(): %v", err)
 	}
 
-	newTr, err = clients.taskrunInformer.Lister().TaskRuns(taskRun.Namespace).Get(taskRun.Name)
+	newTr, err = clients.Pipeline.PipelineV1alpha1().TaskRuns(taskRun.Namespace).Get(taskRun.Name, metav1.GetOptions{})
 	if err != nil {
 		t.Fatalf("Unexpected error fetching taskrun: %v", err)
 	}
 	if d := cmp.Diff(newTr.Status.GetCondition(duckv1alpha1.ConditionSucceeded), buildSt, ignoreLastTransitionTime); d != "" {
 		t.Fatalf("Taskrun Status diff -want, +got: %v", d)
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: consider using `Errorf` here

**Comment by bobcatfish** [comment]
> @bobcatfish : It would be more useful to propagate meaningful msg in status to explain the failure instead of just setting the status to false. For now it could be the error msg too.

WDYT?

Ooooh that's a great idea @shashwathi , maybe that means I can remove some of the logging too! and not assert on it :D

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 1 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: nice, will do!!

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 2 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: I can do that - can you explain _why_ we need to do this, I just don't quite understand what it's for.

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 3 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: i was going to do this, but it turned out we already had a test case covering it and it seemed to happen anyway! (i guess b/c of the state of the taskruns - but I'll initialize it anyway like you describe, just to be safe!

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go (hunk 4 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: good idea!

**Comment by bobcatfish** [comment]
> @bobcatfish : It would be more useful to propagate meaningful msg in status to explain the failure instead of just setting the status to false. For now it could be the error msg too.
> WDYT?

This is making everything 1000x better, great call @shashwathi !! 😍

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 1 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: oh wait, i think im kind of confused - gonna move this to the other thread in this PR on initialize

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 2 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: So [TaskRunStatus has `InitializeConditions`](https://github.com/knative/build-pipeline/blob/394a25bed41a584880e09f5a1596e190033f186c/pkg/apis/pipeline/v1alpha1/taskrun_types.go#L97), but I'm still not understanding what ends up calling it - can you explain what I should add, where and how it will work?

(I'm guessing add `InitializeConditions` to PipelineRunStatus - and should I update both controllers to call this at the beginning of the reconcile loop?)

**Comment by bobcatfish** [comment]
Ready for another look @shashwathi and @pivotal-nader-ziada but @shashwathi I still haven't added the `Initialize` changes you mentioned, I need a bit of guidance there!

**Comment by pivotal-nader-ziada** [comment]
I think the PR is good to go now, will check if @shashwathi can any issues before lgtm

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 2 above)
Comment (shashwathi) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: I am sorry for doing a crappy job in explaining that. Let me rephrase, `InitializeConditions` function in taskrun checks if ducktype condition is nil and initialize condition defined in `taskrunCondSet` to `Unknown` status. Uses ducktype defined function under the hood.

Ref for function definition: https://github.com/knative/build-pipeline/blob/394a25bed41a584880e09f5a1596e190033f186c/vendor/github.com/knative/pkg/apis/duck/v1alpha1/condition_set.go#L295-L302



> both controllers to call this at the beginning of the reconcile loop?

Yes exactly.

PipelineRun does not have `InitializeConditions`, so you could consider initializing that function which can remove the following block of code.

```
if tr.Status.GetCondition(duckv1alpha1.ConditionSucceeded) == nil {
 tr.Status.SetCondition(&duckv1alpha1.Condition{
 Type:   duckv1alpha1.ConditionSucceeded,
 Status: corev1.ConditionUnknown,
 Reason: ReasonRunning,
 })
 }
```

An example of using `InitializeConditions ` function in serving: 
https://github.com/knative/serving/blob/c3a95a733f195cce81922299059c3acfe27ce30b/pkg/reconciler/v1alpha1/service/service.go#L141

I recommended the change because ducktype considers all these cases and has functions defined already. It would be a good practice to adopt these functions in controllers and not duplicate code.

Comment (shashwathi) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: @bobcatfish : If this is increasing the scope of Pr, feel free to discard the recommendation. Functionality is not affected.

Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: thanks for the detailed explanation @shashwathi ! I'll give it a shot and if I do something wonky let me know, then we could maybe put it in another PR instead but I'll see if we can do it here.
